### DISCUSSION TITLE
Tab Completion Ergonomics

### DISCUSSION THREAD

**Discussion by phistep** [zed-industries/zed#20288] [Feature Requests]
I love zed, but the ergonomics of tab-completion are not what I am used to. It currently works like this

1. type a character `c`
   ```
   c|
   ```
2. the popover appears, the first option is selected
   ```
   c|
   ┌──────────┐
   │[case    ]│
   │ class    │
   │ chr      │
   │ Call     │
   └──────────┘ 
   ```
3. when you hit `<tab>`, the selected item is inserted into the buffer and the popover is closed, the cursor is placed at the end of the inserted text
   ```
   case|
   ```
4. if intead you want to select a different item, you can use `<down>` to change the select 
   ```
   c|
   ┌──────────┐
   │ case     │
   │[class   ]│
   │ chr      │
   │ Call     │
   └──────────┘ 
   ```
5. to select the item, press `<enter>` or `<tab>`. Same as (3)

---

I used to use nvim + [CoC](https://github.com/neoclide/coc.nvim), so don't want to take my fingers of the homerow. Using the arrow keys to navigate the options is cumbersome. CoC does it like this:

1. type a character `c`
   ```
   c|
   ```
2. the popover appears, the first 
   ```
   c|
   ┌──────────┐
   │[case    ]│
   │ class    │
   │ chr      │
   │ Call     │
   └──────────┘ 
   ```
3. when you hit `<tab>`, the selected item is inserted into the buffer and the cursor is placed at the end of the inserted text, **but the popover stays open**, the respective option is selected.
   ```
   case|
   ┌──────────┐
   │[case    ]│
   │ class    │
   │ chr      │
   │ Call     │
   └──────────┘ 
   ```
4. if intead you want to select a different item, you can use `<tab>` to cycle through the options (and `<shift-tab>` to cycle reversely). The old insertion is removed and the newly selected item is inserted, the cursor is placed at the end of the inserted text
   ```
   class|
   ┌──────────┐
   │ case     │
   │[class   ]│
   │ chr      │
   │ Call     │
   └──────────┘ 
   ```
5. upon any other keypress than `<tab>` (for example continuing typing a space), the suggestion is accepted and the popover is closed
   ```
   class |
   ```
6. to cancel the completion, pimply press `<esc>`, then the inserted text is removed and the popover is closed.
   ```
   c|
   ```
   
---
   
Frankly I don't see any downside of this, users of the arrow-key/return camp can still do that, but vim users don't have to leave the homerow. It's much faster and more ergonomic. It feels nicer, since you automatically see what you will be inserting and it integrates so much nicer into the flow of typing.

Is it possible to configure zed to behave this way? Would be open to implement this behaviour or accept a PR? (I'm new to rust, but I would give it a try if there is the possibility to have it accepted)

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support

**Reply by phistep** [reply]
https://discord.com/channels/869392257814519848/1481971070889230436
